### ISSUE TITLE
Add Support for Path-Based Routing in Kamal

### ISSUE THREAD

**Issue by jvkassi** [basecamp/kamal-proxy#48]
Currently, Kamal only supports host-based routing, which limits its flexibility when managing requests on a single domain or multiple applications under the same host. To enhance its functionality and improve usability, I suggest implementing path-based routing for the following reasons:

* Efficiency in Resource Usage:

With path-based routing, multiple services can be deployed under a single host (e.g., example.com/app1, example.com/app2), reducing the need for additional hostnames or subdomains. This is particularly beneficial when using wildcard SSL certificates or when managing applications within a single domain.

* Simplified Application Management:

Path-based routing allows for the separation of different applications or services under distinct paths (e.g., /api, /admin). This simplifies management and reduces complexity, especially when dealing with microservices or multiple frontends that share a domain.

**Comment by kevinmcconnell** [comment]
@jvkassi yes, agreed. We've been talking about adding this, and I have a solution in mind that I'm starting to work on. My current thinking is we'll add a path prefix option that you can use when deploying, either in combination with host-based routing or on its own. If you specify a path prefix, then only requests that match the prefix will be routed to that service. That should be enough to have apps share a domain but split according to `/api`, `/admin`, `/app`, etc.

I expect to have something more concrete to try out for this soon.

**Comment by nuschk** [comment]
My usecase (with Kamal 1.9) is to have a CMS and an app server behind the same Traefik instance, on the same domain. What I do is route the root path (`/`) and all paths with a content prefix (`/c/*`), as well as some CMS related paths (`/_next/*` etc.) to the CMS, and have all other requests fall back to the app server. 

I wonder how that would be able with your proposal. It would probably require either some sort of service priority (which is what Traefik offers) or some sort of negation ("not these paths"). And also, not just path prefixes, but also paths (a distinction Traefik makes as well).

Would be great if that would go into your consideration!

**Comment by stefsava** [comment]
Great news that you are working on it, it is very similar to what I need to install multiple Odoo instances on a single server (or more than one with load balancer).

Specifically, Odoo requires / requests to be routed on port 8069, while /websocket requests are routed on port 8072.

Here is the nginx configuration suggested by the official documentation:
https://www.odoo.com/documentation/18.0/administration/on_premise/deploy.html#id8

I use rails and odoo and I am happy to abandon dokku and embrace kamal as soon as possible.

**Comment by grimesp** [comment]
I would also love to see this. Our use case would be deploying multiple Java web applications as separate docker containers as part of a large ERP system. In the Java world, you _can_ have multiple `war` files in a single Java web server (Tomcat), but each war file is an independent Java web application, they all have different URL paths. We like to keep one war file per Tomcat container so we can upgrade components of the ERP independently. It is also nice for security, keeping them in separate docker containers provides a nice layer of isolation.

**Comment by voberoi** [comment]
Is this feature still in the cards?

Was a lot easier when you were using Traefik because this was baked in and I could use Docker labels.

But now I have to deploy my own proxy (Traefik/Caddy/whatever) with its own settings in addition to apps. This is workable, but annoying/more to keep track of and mess up.

**Comment by nikklavzar** [comment]
I am also waiting on this feature before migrating to kamal v2.

**Comment by mkuklis** [comment]
@voberoi I'm curious did you turn off kamal-proxy completely and setup Traefik via `accessories`?

**Comment by voberoi** [comment]
@mkuklis I didn't. `kamal-proxy` is too usefully integrated into Kamal.

My use case is that I have a stack of monitoring tools that I deploy (prometheus, grafana, dozzle) that are basically configured the same way on a single app server. These are pretty basic app deployments.

I started using accessories to deploy them in Kamal v1.

I use Kamal v2 now. And I just have a separate app where:

- The main app is Caddy. Path-based routing is set up there.
- The app uses `kamal-proxy` at a subdomain (e.g. `monitoring.mydomain.com`)
- I have set up all three services as an accessory, and Caddy's set up to proxy to them.
- They're all on the same network and their names are deterministic, so I use those.

This does the job but there are still annoying things about it --

- I don't want to have to set up a proxy in front of the proxy just so I can benefit from path based routing
- `kamal` doesn't allow a public Docker image for the main app. So instead of using the Caddy public image, I had to create my own that extended the public image and was uploaded to my private registry. There's a ticket about this somewhere but I can't find it.

**Comment by mkuklis** [comment]
@voberoi thank you so much for finding time to respond to me. Your setup makes a lot of sense given this current limitation in kamal-proxy.

**Comment by kevinmcconnell** [comment]
Thanks for all the suggestions on this issue so far, everyone. Apologies for the radio silence :grimacing:

The feature is still on the cards, and it's been on my list to get it shipped. I have some sketches of how I think it should work, and I'm aiming to have a working version soon that folks can try out. Stay tuned :)

**Comment by stefsava** [comment]
As written earlier for my use case ([`Odoo deployment support`](#issuecomment-2443738690)), you need to redirect a specific path to a specific port.

- one.example.com/websocket -> port 8072
- one.example.com -> port 8069
- two.example.com/websocket -> port 8072
- two.example.com -> port 8069

I hope this PR supports this need

**Comment by kevinmcconnell** [comment]
@stefsava if the websocket part is deployed as its own service, then this should be covered. The services could be something along the lines of:

```
kamal-proxy deploy web \
  --host=one.example.com,two.example.com --target={host}:8069

kamal-proxy deploy websocket \
  --host=one.example.com,two.example.com --target={host}:8072 --path-prefix=/websocket
```

If they need to refer to the same deployment, though, then I think that is something else. That may require some other changes in Kamal -- maybe something like support for defining multiple proxy deployments for a single service (such that Kamal would issue something like the above whenever that service is deployed).

**Comment by voberoi** [comment]
Thank you! This looks great!

Will there be a companion PR in kamal to add these settings to the service configs?

**Comment by voberoi** [comment]
Great! And, just to clarify -- these will be possible to set on accessories as well?

**Comment by kevinmcconnell** [comment]
> Great! And, just to clarify -- these will be possible to set on accessories as well?

I think so, yes, because we should be able to put them in that same `proxy` section.

**Comment by voberoi** [comment]
Amazing. Thanks!

**Comment by nuschk** [comment]
Great news, thanks for your work!

For my use case of running a CMS and the actual app on the same domain, I'd need to be able to specify *multiple* path prefixes though. Even better would be to have the option of specifying additional paths (not prefixes) as well.

- `https://mydomain.com/`  => CMS
- `https://mydomain.com/about`  => CMS
- `https://mydomain.com/pricing`  => CMS
- `https://mydomain.com/content/...`  => CMS
- `https://mydomain.com/api/...`  => App server
- `https://mydomain.com/account/login`  => App server
- `https://mydomain.com/listing/983748`  => App server

Yes, technically, I could have an `/app` prefix on all app routes, but due to SEO (e.g. for the last URL), I wouldn't want to do that. Also, I could use a `/cms` prefix for all of the CMS, but then would need to handle `/` separately (currently not doable I guess).

If I could use multiple prefixes, I'd just put all the app prefixes (`/api`, `/account`, `/listing`) in there and serve the rest to the CMS. If I could have paths also, I would server `/`, `/about`, `/pricing` paths, as well as `/content/` prefix to the CMS, the rest would go to the app.

I haven't read the PR's actual code, but how is the priority exactly determined? Do services with path prefixes take precedence over services without? In my scenario, that would make sense.

Thanks!

**Comment by kevinmcconnell** [comment]
@nuschk thanks for that example. Right now that would only be possible if we add support for multiple path prefixes per service. I think it may make sense to do that, but I'd like to first see how common that need is. Still trying to find the right balance between simplicity and flexibility here :)

Re priority -- we do it by prefix length. So services with path prefixes do take precedence over those without, and longer prefixes match before shorter ones. Essentially, for a given host, we sort the services in order of prefix length (longest to shortest/empty), and then the routing will look for the first one of those that matches the request.

Anyway, I'll give some thought to the multiple-path option, and will let you know if/when it gets added. Thanks!

**Comment by aarroisi** [comment]
Hi @kevinmcconnell, thank you for this! How to use this in the kamal confiig? Or is this still not available to be used in kamal?

**Comment by ACPK** [comment]
https://github.com/basecamp/kamal/issues/1458

**Comment by ACPK** [comment]
@kevinmcconnell Should the "In Progress" label be removed as it was merged into main.  

@aarroisi I’ve opened a pull request to add support for this feature in Kamal: https://github.com/basecamp/kamal/pull/1551. It includes notes on how to use it before it’s merged into main.

**Comment by kevinmcconnell** [comment]
@ACPK thanks for the label reminder! I've removed it.

And thanks for your Kamal PR as well, it looks great.

**Comment by Splines** [comment]
I'm wondering if this [path-based routing](https://kamal-deploy.org/docs/configuration/proxy/#path-based-routing) can be used to serve assets or media like `https://example.com/media/served-via-nginx.mp4` via nginx set up as Kamal accessory.
- The [`app_port`](https://kamal-deploy.org/docs/configuration/proxy/#app-port) property seems to only refer to the main web container, not any accessory. So, I can't easily route all the traffic through nginx if I want to keep using the Kamal proxys. But with such a setup we wouldn't get zero-downtime deploy, so it makes sense that `app_port` cannot point to a port served by an accessory.
- The [`asset_path`](https://kamal-deploy.org/docs/configuration/overview/#asset-path) property seems to be only for serving assets _in-between_ deployments, but not during the runtime itself, right?

Can I somehow configure the Kamal proxy to redirect every `/media/` request to the nginx accessory? (Or maybe even let the Kamal proxy serve those assets directly from disk? But nginx is probably better for this job.)
